File: app/frontend/src/api/students.ts
import { api } from './client'
import type { Student, StudentDetail } from './types'

export const studentsApi = {
  list: (archived = false) =>
    api.get<Student[]>('/students', { params: { archived } }).then((r) => r.data),

  get: (id: number) => api.get<StudentDetail>(`/students/${id}`).then((r) => r.data),

  create: (data: {
    name: string
    color: string
    hourly_rate: number
    phone?: string | null
    note?: string | null
  }) => api.post<Student>('/students', data).then((r) => r.data),

  update: (id: number, data: Partial<Student>, recalcMode = 'today') =>
    api
      .patch<{ student: Student; affected_future_lessons: number }>(`/students/${id}`, data, { params: { recalc_mode: recalcMode } })
      .then((r) => r.data),

  archive: (id: number) => api.post<Student>(`/students/${id}/archive`).then((r) => r.data),

  unarchive: (id: number) => api.post<Student>(`/students/${id}/unarchive`).then((r) => r.data),

  remove: (id: number) => api.delete(`/students/${id}`),
}

File: app/backend/app/routers/students.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.deps import get_db
from app.schemas.student import (
    StudentCreate,
    StudentDetail,
    StudentOut,
    StudentUpdate,
)
from app.services import student_service

router = APIRouter(prefix="/students", tags=["students"])


@router.get("", response_model=list[StudentOut])
def list_students(archived: bool | None = False, db: Session = Depends(get_db)):
    return student_service.list_students(db, archived=archived)


@router.post("", response_model=StudentOut)
def create_student(payload: StudentCreate, db: Session = Depends(get_db)):
    s = student_service.create_student(db, payload)
    return StudentOut.model_validate(s)


@router.get("/{student_id}", response_model=StudentDetail)
def get_student(student_id: int, db: Session = Depends(get_db)):
    return student_service.get_student_detail(db, student_id)


@router.patch("/{student_id}")
def update_student(student_id: int, payload: StudentUpdate, recalc_mode: str = "today", db: Session = Depends(get_db)):
    s, affected = student_service.update_student(db, student_id, payload, recalc_mode)
    return {
        "student": StudentOut.model_validate(s).model_dump(),
        "affected_future_lessons": affected,
    }


@router.post("/{student_id}/archive", response_model=StudentOut)
def archive_student(student_id: int, db: Session = Depends(get_db)):
    s = student_service.archive_student(db, student_id, True)
    return StudentOut.model_validate(s)


@router.post("/{student_id}/unarchive", response_model=StudentOut)
def unarchive_student(student_id: int, db: Session = Depends(get_db)):
    s = student_service.archive_student(db, student_id, False)
    return StudentOut.model_validate(s)


@router.delete("/{student_id}", status_code=204)
def delete_student(student_id: int, db: Session = Depends(get_db)):
    student_service.delete_student(db, student_id)
    return None
